### ISSUE TITLE
sknn.mlp.classifier much slower than nolearn.lasagne.neuralnet

### ISSUE THREAD

**Issue by fayeshine** [aigamedev/scikit-neuralnetwork#169]
In mnist bench example,
sknn.mlp.classifier much slower than nolearn.lasagne.neuralnet on the same parameter, however it get higher accuracy
What reason cause this?

**Comment by alexjc** [comment]
I haven't compared recently. Can you paste your scripts, data or results? Are you using the GPU for both?

**Comment by fayeshine** [comment]
1: GPU for both, 2: no convolution, 3: no validation
my hardware is 5960x, 64gb mem, titan X
The below is my script, just modified from the mnist example, besides I cannot print the verbose. I think there's something happened in recent commits in THEANO, please try the bleeding edge version of theano



#%%
# -*- coding: utf-8 -*-
from __future__ import (absolute_import, unicode_literals, print_function)

import sys
import time
import numpy as np

#if len(sys.argv) == 1:
#    print("ERROR: Please specify implementation to benchmark, 'sknn' or 'lasagne'.")
#    sys.exit(-1)

np.set_printoptions(precision=4)
np.set_printoptions(suppress=True)

from sklearn.base import clone
from sklearn.cross_validation import train_test_split
from sklearn.datasets import fetch_mldata
from sklearn.metrics import classification_report
from sklearn import * 

mnist = fetch_mldata('mnist-original')
X_train, X_test, y_train, y_test = train_test_split(
        (mnist.data / 255.0).astype(np.float32),
        mnist.target.astype(np.int32),
        test_size=1.0/7.0, random_state=1234)
        
to_bench=['sknn']
#%%

classifiers = []


if 'sknn' in to_bench:
    from sknn.mlp import Classifier, Layer, Convolution

    clf = Classifier(
        layers=[
            #Convolution("Rectifier", channels=10, pool_shape=(2,2), kernel_shape=(3, 3)),
            Layer('Rectifier', units=200),
            Layer('Softmax')],
        learning_rate=0.02,
        learning_rule='nesterov',
        learning_momentum=0.9,
        batch_size=300,
        valid_size=0.0,
        # valid_set=(X_test, y_test),
        n_stable=10,
        n_iter=10,
        verbose=False)
    classifiers.append(('sknn.mlp', clf))

if 'lasagne' in to_bench:
    from nolearn.lasagne import NeuralNet, BatchIterator
    from lasagne.layers import InputLayer, DenseLayer
    from lasagne.nonlinearities import *
    from lasagne.updates import nesterov_momentum

    clf = NeuralNet(
        layers=[
            ('input', InputLayer),
            ('hidden1', DenseLayer),
            ('output', DenseLayer),
            ],
        input_shape=(None, 784),
        output_num_units=10,
        output_nonlinearity=softmax,
        eval_size=0.0,

        more_params=dict(
            hidden1_num_units=200,
            hidden1_nonlinearity=rectify
        ),

        update=nesterov_momentum,
        update_learning_rate=0.02,
        update_momentum=0.9,
        batch_iterator_train=BatchIterator(batch_size=300),

        max_epochs=10,
        verbose=1)
    classifiers.append(('nolearn.lasagne', clf))


RUNS = 1

for name, orig in classifiers:
    times = []
    accuracies = []
    for i in range(RUNS):
        start = time.time()

        clf = clone(orig)
        clf.random_state = int(time.time())
        clf.fit(X_train, y_train)

        accuracies.append(clf.score(X_test, y_test))
        times.append(time.time() - start)

    a_t = np.array(times)
    a_s = np.array(accuracies)

    y_pred = clf.predict(X_test)

    print("\n"+name)
    print("\tAccuracy: %5.2f%% ±%4.2f" % (100.0 * a_s.mean(), 100.0 * a_s.std()))
    print("\tTimes:    %5.2fs ±%4.2f" % (a_t.mean(), a_t.std()))
    print("\tReport:")
    print(classification_report(y_test, y_pred))

**Comment by alexjc** [comment]
Thanks, I'll see what I can do. If you have any further insights about recent changes in Theano, let me know!

**Comment by alexjc** [comment]
Oh, oops! So it's iterating `batch_size` more times, which explains better quality. Code just changed to fix dense iterators. Adding a test for this!

**Comment by alexjc** [comment]
Fixed in #170. Looking really much faster now...

```
nolearn.lasagne
        Accuracy: 96.58% ±0.10
        Times:    54.66s ±2.26

sknn.mlp
        Accuracy: 96.64% ±0.08
        Times:    18.42s ±11.25
```

**Comment by fayeshine** [comment]
Thanks a lot! I'll test on my dataset and give you feedback.

**Comment by alexjc** [comment]
It's still slower, depending on the format of your data, as mentioned in the PR. Going to look into the differences.
